### ISSUE TITLE
Add struct-level annotation for renaming fields

### ISSUE THREAD

**Issue by erickt** [serde-rs/serde#140]
Serde supports field level renaming with:

``` rust
struct Foo {
  #[serde(rename="baz")]
  bar: usize
}
```

But this can be inconvenient to use if you have a lot of fields you need to rename, such as going from rust's "snake-case" style into "camel-case". It would be nice to have struct level renaming support for this:

``` rust
#[serde(rename_camel_case)]
#[derive(Serialize, Deserialize)]
struct CamelCase {
  foo_bar_baz: usize
}

#[serde(rename_underscore_to_hypens)]
#[derive(Serialize, Deserialize)]
struct Hyphens {
  foo_bar_baz: usize
}

```

Would get serialized to `fooBarBaz`, and `foo-bar-baz`.

**Comment by erickt** [comment]
This was implemented a while ago.

**Comment by williamho** [comment]
does this feature still exist? I can't seem to find it in the docs.

**Comment by scottpleb** [comment]
@williamho figuring out how to do this was surprisingly tricky, as the syntax apparently changed a few times, and it really doesn't seem to be documented anywhere. This is what worked for me:
```rust
#[derive(Debug, Deserialize)]
pub struct UploadResponse {
    fid: String,
    #[serde(rename="fileUrl")]
    file_url: String
}
```

**Comment by dtolnay** [comment]
@scottpleb your specific use case is documented here: https://serde.rs/attr-rename.html

All the attributes are documented here: https://serde.rs/attributes.html

**Comment by dtolnay** [comment]
I don't think this was ever implemented.

**Comment by dtolnay** [comment]
@46bit mentioned possibly being interested in working on this.

Possibly the easiest implementation would be [right above this line](https://github.com/serde-rs/serde/blob/v0.9.8/serde_codegen_internals/src/ast.rs#L49), traverse the "body" and apply all of the renames in place. So if we have the item attrs asking for camelCase, we would change all of the names of the fields (except those that already have a rename) to be camelCase.

Alternatively you could take a reference to [this guy](https://github.com/serde-rs/serde/blob/v0.9.8/serde_codegen_internals/src/ast.rs#L39) or just its bulk rename field, and pass it through [these<sup>1</sup>](https://github.com/serde-rs/serde/blob/v0.9.8/serde_codegen_internals/src/ast.rs#L42) [two<sup>2</sup>](https://github.com/serde-rs/serde/blob/v0.9.8/serde_codegen_internals/src/ast.rs#L44) calls so the field names get set correctly right away rather than having to update them later. Try whichever one seems easier to you.

In terms of the actual attributes, I don't like @erickt's `serde(rename_camel_case)` and `serde(rename_underscore_to_hypens)`, would prefer to use a string so the names can look evocative like what the reader would see if these were individual rename attributes. Something like:

```rust
#[serde(rename_all = "snake_case")]
#[serde(rename_all = "kebab-case")]
#[serde(rename_all = "SNAKE_CASE")]
#[serde(rename_all = "camelCase")]
#[serde(rename_all = "PascalCase")]
```

Bonus points for not even looking at the word, just the casing.

```rust
#[serde(rename_all = "easterEggCase")]
```

**Comment by 46bit** [comment]
Awesome. I like `rename_all` and agree exactly on the code demos you've put. But I'd like to allow users to use their own renaming function, similar to with `#[serde(default = "fn")]`. 

The obvious signature would be a mapping from field name to output name, `fn(&str) -> &str`. We could also provide the other field names in this container with `fn(&str, [&str]) -> &str` - but this doesn't seem worthwhile.

**Comment by 46bit** [comment]
Normal renaming is done in `Variant::from_ast`/`Field::from_ast` so we don't know which fields are renamed when we get to `Item::from_ast`. I add a boolean flag to `Variant` and `Field` that indicates that renaming has happened.

I note there's `#[serde(rename(serialize = "name"))]` direction-specific renaming. I'm not inclined to implement that for `rename_all` as it seems rather special-purpose.

**Comment by dtolnay** [comment]
Implemented in #788.
